### ISSUE TITLE
Conformer not working compared to Transducer during streaming decode

### ISSUE THREAD

**Issue by stefan-falk** [espnet/espnet#2540]
So, I have implemented the vanilla RNN-T (Transducer) model in Tensorflow and also a `greedy_decode()` function similar to

https://github.com/espnet/espnet/blob/88bb9dba83636c4bf12f10a870a7d8c5f9515b9f/espnet/nets/beam_search_transducer.py#L36

It works fine on the Transducer but not for the Conformer. I noticed that the Conformer needs much larger audio chunks to be presented in order to produce _some_ sensible output but the overall predictions are still (very) bad. 

In the original Conformer paper [1] they do not explain which method has been used exactly for decoding and how streaming works. For the Transducer, my greedy_decode() works pretty well but the Conformer, as I said, does not really give sensible results (unless I present rather huge audio chunks).

When it comes to streaming, does the approach work differently for the Transducer and Conformer?

Is there a chance that masking is required specifically for the Self-Attention mechanism? My Conformer gets trained exactly like the Transducer. Could that be the issue here?

Another thing that might be the issue: The Transducer gets encoder state from one chunk to the next which allows the acoustic model to continue encoding from where it previously left off. Since the Conformer has no internal state, how does the Encoder layer "remember" what it saw in the previous chunk?

Thank you for any help on this.

[1] https://arxiv.org/pdf/2005.08100.pdf

**Comment by stefan-falk** [comment]
@pzelasko Any chance you know something about this topic?

I just saw this comment:

https://github.com/espnet/espnet/blob/64ca9fa27e63e5b28792d1156a0d19159b738733/espnet/nets/pytorch_backend/streaming/window.py#L4

Do we need to stream audio chunks in a different way for the Conformer?

**Comment by stefan-falk** [comment]
Okay, it seems that I am the victim of a misunderstanding here. It seems that the Conformer cannot be used in a streaming environment _because_ of the self-attention.

**Comment by joespitch** [comment]
If you are looking for ideas about streaming transformers and conformers, this is probably the best online implementation resource currently available: https://github.com/cywang97/StreamingTransformer

**Comment by AdolfVonKleist** [comment]
If you are looking for ideas about streaming transformers this is probably the best online implementation resource at the moment:  https://github.com/cywang97/StreamingTransformer

**Comment by Peidong-Wang** [comment]
I think conformer has not been used for streaming ASR yet. Both ContextNet and conformer benefits from the global information in the whole utterance. The Emformer work from FAIAR may be another thread to follow for streaming ASR, but the code may not be publically available.

**Comment by AdolfVonKleist** [comment]
It looks like the concept has in fact been applied here for the conformer: https://github.com/cywang97/StreamingTransformer/issues/7 ; whether that's an optimal choice for streaming/implementation I guess is debatable. The performance degrades compared to offline for the reason you describe.
